browser/ui/webui/brave_education/education_page_ui_browsertest.cc

```diff
@@ -0,0 +1,147 @@
+#include <utility>
+
+#include "base/memory/weak_ptr.h"
+#include "brave/components/brave_education/common/education_urls.h"
+#include "brave/components/brave_vpn/common/buildflags/buildflags.h"
+#include "brave/components/constants/webui_url_constants.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/common/webui_url_constants.h"
+#include "chrome/test/base/chrome_test_utils.h"
+#include "chrome/test/base/in_process_browser_test.h"
+#include "chrome/test/base/ui_test_utils.h"
+#include "content/public/browser/web_ui_data_source.h"
+#include "content/public/test/browser_test.h"
+#include "content/public/test/browser_test_utils.h"
+
+namespace brave_education {
+
+class EducationPageUIBrowserTest : public InProcessBrowserTest {
+ protected:
+  void SetUpOnMainThread() override { CreateAndAddWebUITestDataSource(); }
+
+  void NavigateToEducationPage(EducationPageType page_type) {
+    auto webui_url = GetEducationPageBrowserURL(page_type);
+    ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), webui_url));
+    auto* web_contents = chrome_test_utils::GetActiveWebContents(this);
+    ASSERT_TRUE(WaitForLoadStop(web_contents));
+  }
+
+  void PostMessageFromIFrame(std::string_view message_data) {
+    message_data_ = message_data;
+    auto* web_contents = chrome_test_utils::GetActiveWebContents(this);
+    ASSERT_TRUE(content::ExecJs(web_contents, R"(
+      const iframe = document.getElementById('content')
+      iframe.src = "chrome://webui-test/"
+    )"));
+  }
+
+ private:
+  void CreateAndAddWebUITestDataSource() {
+    auto* web_contents = chrome_test_utils::GetActiveWebContents(this);
+
+    content::WebUIDataSource* source = content::WebUIDataSource::CreateAndAdd(
+        web_contents->GetBrowserContext(), chrome::kChromeUIWebUITestHost);
+
+    source->OverrideContentSecurityPolicy(
+        network::mojom::CSPDirectiveName::FrameAncestors,
+        "frame-ancestors chrome://* 'self';");
+
+    source->SetRequestFilter(
+        base::BindRepeating([](const std::string&) { return true; }),
+        base::BindRepeating(&EducationPageUIBrowserTest::HandleWebUITestRequest,
+                            weak_factory_.GetWeakPtr()));
+  }
+
+  void HandleWebUITestRequest(
+      const std::string& path,
+      content::WebUIDataSource::GotDataCallback callback) {
+    if (path == "post-message.js") {
+      std::string js = "window.parent.postMessage(" + message_data_ + ", '*')";
+      std::move(callback).Run(
+          base::MakeRefCounted<base::RefCountedString>(std::move(js)));
+      return;
+    }
+
+    static const char kHTML[] = R"(
+      <!doctype html>
+      <html>
+      <script src='/post-message.js'></script>
+      <body>
+        Hello world!
+      </body>
+      </html>
+    )";
+
+    std::move(callback).Run(
+        base::MakeRefCounted<base::RefCountedString>(std::move(kHTML)));
+  }
+
+  std::string message_data_;
+  base::WeakPtrFactory<EducationPageUIBrowserTest> weak_factory_{this};
+};
+
```

Comment:
you need to test failure cases for these as well. For instance what happens when these are called from profiles where rewards, vpn, aichat etc... are not allowed? Also when vpn is disabled by buildflag.

Reply:
These cases are generally tested in `education_page_handler_unittest`.